Dapper/SqlMapper.GridReader.Async.cs

```diff
@@ -94,6 +121,42 @@ private Task<IEnumerable<T>> ReadAsyncImpl<T>(Type type, bool buffered)
                 }
             }
 
+            private Task<T> ReadFirstOrDefaultAsyncImpl<T>(Type type)
+            {
+                var dbReader = reader as DbDataReader;
+                if (dbReader != null) return ReadFirstOrDefaultAsyncImplImpl<T>(dbReader, type);
+
+                // no async API available; use non-async and fake it
+                return Task.FromResult<T>(ReadFirstOrDefaultImpl<T>(type));

```

Comment:
Since the reader is null going into `FromResult`, that code path is strictly for throwing the error I assumed - that's why I wanted to confirm your intent here vs. throwing. I'm cool with it - it should be the exceptional/rare case anyway and makes sense, but worth a sanity check is all.